service/worker/scheduler/workflow.go

```diff
@@ -246,12 +250,39 @@ func (s *scheduler) run() error {
 		for s.processBuffer() {
 		}
 		s.updateMemoAndSearchAttributes()
+
+		// if schedule is not paused and out of actions or do not have anything scheduled, delete the schedule workflow after retention period is passed
+		if s.tweakables.RetentionTime > 0 && (!s.Schedule.State.Paused && (nextWakeup.IsZero() || !s.canTakeScheduledAction(false, false))) {
+			// if schedule is empty use the previousProcessedTime as last workflow start time
+			lastWorkflowStarTime := s.now().Sub(*previousProcessedTime)
+			if len(s.Info.RecentActions) > 0 {
+				lastWorkflowStarTime = s.now().Sub(*s.Info.RecentActions[len(s.Info.RecentActions)-1].ScheduleTime)
+			}
+
+			if lastWorkflowStarTime >= s.tweakables.RetentionTime {
+				activityOptions := workflow.ActivityOptions{
+					ScheduleToCloseTimeout: 1 * time.Hour,
+					StartToCloseTimeout:    5 * time.Second,
+					RetryPolicy: &temporal.RetryPolicy{
+						InitialInterval: 1 * time.Second,
+						MaximumInterval: 60 * time.Second,
+					},
+				}
+				ctx := workflow.WithActivityOptions(s.ctx, activityOptions)
+				err := workflow.ExecuteActivity(ctx, s.a.DeleteScheduleWorkflow, s.State.ScheduleId).Get(s.ctx, nil)
+				if err != nil {
+					s.logger.Error("Failed to delete schedule workflow", "err", err)
```

Comment:
Can we just exit this schedule workflow and let it stay as completed for namespace's retention time. Directly delete workflow is not normal. We use ContinueAsNew already which means there are already completed workflows, so this would be similar to those closed ones.

Reply:
Sounds good, I'll make a new PR and add make it simpler.

Reply:
Do we want to keep the retention period to 7 days? I think that might be a little bit too long will lead to 7 days + workflow retention (my understanding is that is also 7 days) before it gets deleted. Should be shorten the retention period here?